File: android/ac-core/src/main/kotlin/com/microsoft/adaptivecards/core/AdaptiveCards.kt
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

package com.microsoft.adaptivecards.core

import android.util.Log
import com.microsoft.adaptivecards.core.caching.CardCache
import com.microsoft.adaptivecards.core.models.AdaptiveCard
import com.microsoft.adaptivecards.core.models.CardElement
import com.microsoft.adaptivecards.core.models.CardInput
import com.microsoft.adaptivecards.core.models.Container
import com.microsoft.adaptivecards.core.models.ColumnSet
import com.microsoft.adaptivecards.core.models.Carousel
import com.microsoft.adaptivecards.core.models.TabSet
import com.microsoft.adaptivecards.core.models.UnknownElement
import com.microsoft.adaptivecards.core.parsing.CardParser

/**
 * Primary entry point for the Adaptive Cards SDK.
 * Provides standalone parsing APIs that can be used without a view.
 *
 * ```kotlin
 * // Simple parse
 * val result = AdaptiveCards.parse(jsonString)
 * if (result.isValid) {
 *     val card = result.card!!
 * }
 *
 * // Pre-parse for caching (e.g., in a RecyclerView prefetch)
 * val results = jsonStrings.map { AdaptiveCards.parse(it) }
 * ```
 */
object AdaptiveCards {

    private const val TAG = "AdaptiveCards"

    /**
     * Parse an Adaptive Card JSON string into a card model.
     *
     * Results are cached via [CardCache.shared]. Calling `parse` with the same JSON
     * string returns a cached result with `parseTimeMs == 0`.
     *
     * @param json A JSON string representing an Adaptive Card
     * @return A [ParseResult] containing the card, warnings, and timing info
     */
    fun parse(json: String): ParseResult = parse(json, CardCache.shared)

    /**
     * Parse with a specific cache instance (or null to disable caching).
     */
    fun parse(json: String, cache: CardCache?): ParseResult {
        if (json.isEmpty()) {
            return ParseResult(error = ParseError.Empty())
        }

        // Check cache
        if (cache != null) {
            val cached = cache.cachedCard(json)
            if (cached != null) {
                return ParseResult(
                    card = cached.first,
                    warnings = cached.second,
                    parseTimeMs = 0.0,
                    cacheHit = true
                )
            }
        }

        // Parse
        val startNs = System.nanoTime()
        return try {
            val card = CardParser.parse(json)
            val parseTimeMs = (System.nanoTime() - startNs) / 1_000_000.0
            val warnings = collectWarnings(card)

            // Cache the result
            cache?.cacheCard(card, warnings, json)

            ParseResult(
                card = card,
                warnings = warnings,
                parseTimeMs = parseTimeMs,
                cacheHit = false
            )
        } catch (e: Exception) {
            val parseTimeMs = (System.nanoTime() - startNs) / 1_000_000.0
            Log.e(TAG, "Failed to parse card", e)
            ParseResult(
                error = ParseError.DecodingFailed(e.message ?: "Unknown error"),
                parseTimeMs = parseTimeMs
            )
        }
    }

    /**
     * Clears the shared cache. Call on low-memory events.
     */
    fun clearCache() {
        CardCache.shared.clearAll()
    }

    // MARK: - Prefetch

    /**
     * Pre-parse cards that are about to scroll into view.
     * Call from RecyclerView.Adapter.onBindViewHolder or LazyColumn prefetch.
     */
    fun prefetch(jsons: List<String>, configuration: CardConfiguration = CardConfiguration.Default) {
        Thread {
            for (json in jsons) {
                parse(json, configuration.cache)
            }
        }.start()
    }

    /**
     * Pre-warm image caches for pre-parsed cards.
     */
    @JvmName("prefetchCards")
    fun prefetch(cards: List<AdaptiveCard>, configuration: CardConfiguration = CardConfiguration.Default) {
        // Future: extract image URLs and pre-warm image cache
    }

    /**
     * Cancel prefetch for cards that scrolled out of range.
     */
    fun cancelPrefetch(jsons: List<String>) {
        // Future: cancel in-flight image prefetch tasks
    }

    /**
     * Walk the parsed card tree and collect warnings for unknown types, etc.
     */
    private fun collectWarnings(card: AdaptiveCard): List<ParseWarning> {
        val warnings = mutableListOf<ParseWarning>()

        fun walk(elements: List<CardElement>?) {
            elements ?: return
            for (element in elements) {
                when (element) {
                    is UnknownElement -> {
                        val typeName = element.unknownType ?: "Unknown"
                        warnings.add(
                            ParseWarning(
                                code = ParseWarning.Code.UNKNOWN_ELEMENT_TYPE,
                                message = "Unknown element type: $typeName"
                            )
                        )
                    }
                    is CardInput -> {
                        if (element.id.isNullOrBlank()) {
                            warnings.add(
                                ParseWarning(
                                    code = ParseWarning.Code.MISSING_INPUT_ID,
                                    message = "Input element of type '${element.type}' is missing required 'id' property"
                                )
                            )
                        }
                    }
                    else -> {}
                }

                // Recurse into containers
                when (element) {
                    is Container -> walk(element.items)
                    is ColumnSet -> element.columns?.forEach { walk(it.items) }
                    is Carousel -> element.pages.forEach { walk(it.items) }
                    is TabSet -> element.tabs.forEach { walk(it.items) }
                    else -> {}
                }
            }
        }

        walk(card.body)
        return warnings
    }
}

File: ios/Sources/ACCore/AdaptiveCards.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

import Foundation

/// Primary entry point for the Adaptive Cards SDK.
/// Provides standalone parsing APIs that can be used without a view.
///
/// ```swift
/// // Simple parse
/// let result = AdaptiveCards.parse(jsonString)
/// if let card = result.card {
///     // Use the card
/// }
///
/// // Pre-parse for caching (e.g., in a list prefetch)
/// let results = jsonStrings.map { AdaptiveCards.parse($0) }
/// ```
public enum AdaptiveCards {

    // MARK: - Parsing

    /// Parse an Adaptive Card JSON string into a card model.
    ///
    /// Results are cached via `CardCache.shared`. Calling `parse` with the same JSON
    /// string returns a cached result with `parseTimeMs == 0`.
    ///
    /// - Parameter json: A JSON string representing an Adaptive Card
    /// - Returns: A `ParseResult` containing the card, warnings, and timing info
    public static func parse(_ json: String) -> ParseResult {
        return parse(json, cache: CardCache.shared)
    }

    /// Parse with a specific cache instance.
    public static func parse(_ json: String, cache: CardCache?) -> ParseResult {
        guard !json.isEmpty else {
            return ParseResult(error: .empty)
        }

        // Check cache
        if let cache = cache, let cached = cache.cachedCard(for: json) {
            return ParseResult(
                card: cached.card,
                warnings: cached.warnings,
                parseTimeMs: 0,
                cacheHit: true
            )
        }

        // Parse
        let start = CFAbsoluteTimeGetCurrent()
        let parser = CardParser()

        do {
            let card = try parser.parse(json)
            let parseTimeMs = (CFAbsoluteTimeGetCurrent() - start) * 1000
            let warnings = collectWarnings(for: card)

            // Cache the result
            cache?.cacheCard(card, warnings: warnings, for: json)

            return ParseResult(
                card: card,
                warnings: warnings,
                parseTimeMs: parseTimeMs,
                cacheHit: false
            )
        } catch let error as CardParserError {
            let parseTimeMs = (CFAbsoluteTimeGetCurrent() - start) * 1000
            let parseError: ParseError
            switch error {
            case .invalidJSON(let msg):
                parseError = .invalidJSON(msg)
            case .decodingFailed(let msg):
                parseError = .decodingFailed(msg)
            case .missingRequiredField(let msg):
                parseError = .decodingFailed("Missing required field: \(msg)")
            }
            return ParseResult(error: parseError, parseTimeMs: parseTimeMs)
        } catch {
            let parseTimeMs = (CFAbsoluteTimeGetCurrent() - start) * 1000
            return ParseResult(
                error: .decodingFailed(error.localizedDescription),
                parseTimeMs: parseTimeMs
            )
        }
    }

    /// Clears the shared parse cache. Call when receiving memory warnings.
    public static func clearCache() {
        CardCache.shared.clearAll()
    }

    // MARK: - Private Helpers

    /// Walk the parsed card tree and collect warnings for unknown types, inputs without IDs, etc.
    private static func collectWarnings(for card: AdaptiveCard) -> [ParseWarning] {
        var warnings: [ParseWarning] = []

        func walk(_ elements: [CardElement]?) {
            guard let elements = elements else { return }
            for element in elements {
                switch element {
                case .unknown(let type):
                    warnings.append(ParseWarning(
                        code: .unknownElementType,
                        message: "Unknown element type: \(type)"
                    ))
                case .container(let c):
                    walk(c.items)
                case .columnSet(let cs):
                    for column in cs.columns {
                        walk(column.items)
                    }
                case .carousel(let c):
                    for page in c.pages {
                        walk(page.items)
                    }
                case .tabSet(let ts):
                    for tab in ts.tabs {
                        walk(tab.items)
                    }
                default:
                    break
                }

                // Check inputs missing IDs
                if let id = element.elementId, id.isEmpty {
                    warnings.append(ParseWarning(
                        code: .missingInputId,
                        message: "Input element missing ID"
                    ))
                }
            }
        }

        walk(card.body)
        return warnings
    }
}
